### PR TITLE
Migrate DashcardCardParameterMapper to TS

### PR DESCRIPTION
**Pull request by uladzimirdev**
### Description

Migrate DashcardCardParameterMapper to TS

blocked by https://github.com/metabase/metabase/pull/38545

### How to verify

Tests should pass, verify types

### Checklist

- [x] Tests have been added/updated to cover changes in this PR

### CODE REVIEW

File: frontend/src/metabase/dashboard/components/DashCard/DashCardParameterMapper/DashCardCardParameterMapper.styled.tsx [hunk 1]
```diff
@@ -27,7 +27,6 @@ export const TextCardDefault = styled.div`
 export const CardLabel = styled.div`
   font-size: 0.83em;
   margin-bottom: ${space(1)};
-  text-weight: bold;
```
Comment (uladzimirdev) [inline frontend/src/metabase/dashboard/components/DashCard/DashCardParameterMapper/DashCardCardParameterMapper.styled.tsx:30]: it should have been `font-weight`

File: frontend/src/metabase/dashboard/components/DashCard/DashCardParameterMapper/DashCardCardParameterMapper.tsx [hunk 2]
```diff
@@ -72,17 +92,22 @@ const mapDispatchToProps = {
...
+  setParameterMapping: (
+    parameterId: ParameterId,
+    dashcardId: DashCardId,
+    cardId: CardId,
+    target: ParameterTarget | null,
+  ) => void;
+  isMobile: boolean;
+  // virtual cards will not have question
+  question?: Question;
+  mappingOptions: any[];
```
Comment (kulyk) [inline frontend/src/metabase/dashboard/components/DashCard/DashCardParameterMapper/DashCardCardParameterMapper.tsx:109]: any chance we could avoid this `any`?

File: frontend/src/metabase/parameters/utils/mapping-options.ts [hunk 3]
```diff
@@ -61,37 +68,28 @@ function buildTextTagOption(tagName) {
...
+  question: Question,
+  parameter: Parameter | null | undefined = null,
+  card: Card,
+  dashcard: BaseDashboardCard | null | undefined = null,
 ) {
-  if (dashcard && ["heading", "text"].includes(card.display)) {
+  if (dashcard && isVirtualDashCard(dashcard)) {
     const tagNames = tag_names(dashcard.visualization_settings.text || "");
     return tagNames ? tagNames.map(buildTextTagOption) : [];
   }
```
Comment (kulyk) [inline frontend/src/metabase/parameters/utils/mapping-options.ts:80]: IIRC tag names are only available for `heading` and `text` dashcards

**Review by kulyk — APPROVED** [review]

File: frontend/src/metabase/dashboard/components/DashCard/DashCardParameterMapper/DashCardCardParameterMapper.tsx (hunk 2 above)
Comment (uladzimirdev) [reply frontend/src/metabase/dashboard/components/DashCard/DashCardParameterMapper/DashCardCardParameterMapper.tsx:109]: oh, right 😄 will do

File: frontend/src/metabase/parameters/utils/mapping-options.ts (hunk 3 above)
Comment (uladzimirdev) [reply frontend/src/metabase/parameters/utils/mapping-options.ts:80]: right, I'll add this check back.

`type VirtualCardDisplay = "text" | "action" | "link" | "heading"`

Comment (uladzimirdev) [reply frontend/src/metabase/parameters/utils/mapping-options.ts]: [da7ab9e](https://github.com/metabase/metabase/pull/38596/commits/da7ab9ea3336a83768116aa339d26f1f132acfa5)

File: frontend/src/metabase/dashboard/components/DashCard/DashCardParameterMapper/DashCardCardParameterMapper.tsx (hunk 2 above)
Comment (uladzimirdev) [reply frontend/src/metabase/dashboard/components/DashCard/DashCardParameterMapper/DashCardCardParameterMapper.tsx]: [1421ca9](https://github.com/metabase/metabase/pull/38596/commits/1421ca98b1869665984e6ad9a0a7571b539bd3ac)

**Review by kulyk — APPROVED** [review]

File: frontend/src/metabase/parameters/utils/mapping-options.ts [hunk 4]
```diff
@@ -28,13 +50,17 @@ function buildStructuredQuerySectionOptions(query, stageIndex, group) {
       sectionName: getColumnGroupName(groupInfo),
       name: columnInfo.displayName,
       icon: getColumnIcon(column),
-      target: buildColumnTarget(query, stageIndex, column),
+      target: buildColumnTarget(
+        query,
+        stageIndex,
+        column,
+      ) as ParameterDimensionTarget,
```
Comment (kamilmielnik) [inline frontend/src/metabase/parameters/utils/mapping-options.ts:57]: I think this cast hides some typing problems.
We should explicitly specify the return type of `buildColumnTarget` to be `ParameterDimensionTarget` and go from there.

File: frontend/src/metabase/parameters/utils/mapping-options.ts [hunk 5]
```diff
@@ -109,26 +158,28 @@ export function getParameterMappingOptions(
...
   } else {
     const legacyQuery = question.legacyQuery();
     options.push(
-      ...legacyQuery
+      ...(legacyQuery
         .variables(
           parameter ? variableFilterForParameter(parameter) : undefined,
         )
-        .map(buildVariableOption),
+        .map(buildVariableOption) as NativeParameterMappingOptions[]),
```
Comment (kamilmielnik) [inline frontend/src/metabase/parameters/utils/mapping-options.ts:172]: Let's explicitly define return type on `buildVariableOption`.

File: frontend/src/metabase/parameters/utils/mapping-options.ts [hunk 6]
```diff
@@ -109,26 +158,28 @@ export function getParameterMappingOptions(
...
+      ...(legacyQuery
         .dimensionOptions(
           parameter ? dimensionFilterForParameter(parameter) : undefined,
-          parameter ? getTagOperatorFilterForParameter(parameter) : undefined,
         )
         .sections()
-        .flatMap(section => buildNativeQuerySectionOptions(section)),
+        .flatMap(section =>
+          buildNativeQuerySectionOptions(section),
+        ) as NativeParameterMappingOptions[]),
```
Comment (kamilmielnik) [inline frontend/src/metabase/parameters/utils/mapping-options.ts:182]: Let's explicitly define return type on `buildNativeQuerySectionOptions`.

**Review by kamilmielnik — COMMENTED** [review]
The casts hide some underlying typing issues.

File: frontend/src/metabase/parameters/utils/mapping-options.ts (hunk 4 above)
Comment (uladzimirdev) [reply frontend/src/metabase/parameters/utils/mapping-options.ts:57]: that was my initial thought, but it didn't work out. Let me double check

File: frontend/src/metabase-lib/parameters/utils/targets.ts [hunk 7]
```diff
@@ -37,20 +43,30 @@ export function getParameterTargetField(
...
 
-export function buildDimensionTarget(dimension: Dimension) {
-  return ["dimension", dimension.mbql()];
+export function buildDimensionTarget(
+  dimension: Dimension,
+): NativeParameterDimensionTarget {
+  return [
+    "dimension",
+    dimension.mbql() as DimensionReference,
+  ] as NativeParameterDimensionTarget;
```
Comment (kamilmielnik) [inline frontend/src/metabase-lib/parameters/utils/targets.ts:52]: We can drop the cast if we change the return type of `Dimension.prototype.mbql` to `DimensionReference | null | undefined`

```suggestion
  return ["dimension", dimension.mbql()];
```

File: frontend/src/metabase-lib/parameters/utils/targets.ts [hunk 8]
```diff
@@ -37,20 +43,30 @@ export function getParameterTargetField(
...
   query: Lib.Query,
   stageIndex: number,
   column: Lib.ColumnMetadata,
-) {
-  return ["dimension", Lib.legacyRef(query, stageIndex, column)];
+): StructuredParameterDimensionTarget {
+  return [
+    "dimension",
+    Lib.legacyRef(query, stageIndex, column),
+  ] as StructuredParameterDimensionTarget;
```
Comment (kamilmielnik) [inline frontend/src/metabase-lib/parameters/utils/targets.ts:63]: We can drop the cast if we introduce a type guard `const isLocalFieldReference = (value: FieldReference): value is LocalFieldReference`:


```suggestion
  const fieldReference = Lib.legacyRef(query, stageIndex, column);

  if (!isLocalFieldReference(fieldReference)) {
    throw new Error("...");
  }

  return ["dimension", fieldReference];
```

There might be a better way, but I could not find it.

File: frontend/src/metabase-lib/parameters/utils/targets.ts [hunk 9]
```diff
@@ -37,20 +43,30 @@ export function getParameterTargetField(
...
 }
 
-export function buildTemplateTagVariableTarget(variable: TemplateTagVariable) {
-  return ["variable", variable.mbql()];
+export function buildTemplateTagVariableTarget(
+  variable: TemplateTagVariable,
+): ParameterVariableTarget {
+  return ["variable", variable.mbql()] as ParameterVariableTarget;
```
Comment (kamilmielnik) [inline frontend/src/metabase-lib/parameters/utils/targets.ts:69]: We can drop the cast if we explicitly specify the return type of `TemplateTagVariable.prototype.mbql` to be `VariableTarget`:

```suggestion
  return ["variable", variable.mbql()];
```

File: frontend/src/metabase-lib/parameters/utils/targets.ts (hunk 7 above)
Comment (kamilmielnik) [inline frontend/src/metabase-lib/parameters/utils/targets.ts:52]: We can drop the cast if we introduce a type guard `const isVariableTarget = (value: unknown): value is VariableTarget`:

```suggestion
  const variableTarget = dimension.mbql();

  if (!isVariableTarget(variableTarget)) {
    throw new Error("...");
  }

  return ["dimension", variableTarget];
```

Alternatively, we can explicitly specify the return type of `TemplateTagDimension.prototype.mbql` to be `VariableTarget` and then do this:

```suggestion
  if (!(dimension instanceof TemplateTagDimension)) {
    throw new Error("...");
  }
  
  return ["dimension", dimension.mbql()];
```

Or, as a 3rd option, we could change the argument type from `Dimension` to `TemplateTagDimension`, explicitly specify the return type of `TemplateTagDimension.prototype.mbql` to be `VariableTarget` and update usages of `buildDimensionTarget` to make sure they're only ever called with `TemplateTagDimension` (e.g. by filtering `section.items` in `buildNativeQuerySectionOptions`).

Comment (uladzimirdev) [reply frontend/src/metabase-lib/parameters/utils/targets.ts]: [a2d99a1](https://github.com/metabase/metabase/pull/38596/commits/a2d99a16c11199027587bd8f7f1083f0a0c8e17e)

File: frontend/src/metabase-lib/parameters/utils/targets.ts (hunk 9 above)
Comment (uladzimirdev) [reply frontend/src/metabase-lib/parameters/utils/targets.ts]: [a2d99a1](https://github.com/metabase/metabase/pull/38596/commits/a2d99a16c11199027587bd8f7f1083f0a0c8e17e)

File: frontend/src/metabase-lib/parameters/utils/targets.ts (hunk 8 above)
Comment (uladzimirdev) [reply frontend/src/metabase-lib/parameters/utils/targets.ts]: [a2d99a1](https://github.com/metabase/metabase/pull/38596/commits/a2d99a16c11199027587bd8f7f1083f0a0c8e17e)

**Review by kamilmielnik — APPROVED** [review]
Nice :+1: 

ffcd8e42f31c8a6c

File: frontend/src/metabase-lib/parameters/utils/targets.ts [hunk 10]
```diff
@@ -1,14 +1,38 @@
...
+  target: ParameterTarget,
+): target is ParameterVariableTarget {
+  return target[0] === "variable";
+}
+
+function isConcreteFieldReference(
+  reference: FieldReference,
+): reference is LocalFieldReference {
+  return reference[0] === "field";
+}
```
Comment (kulyk) [inline frontend/src/metabase-lib/parameters/utils/targets.ts:30]: `LocalFieldReference` must have a numeric ID. There are field references that use a column name when the ID isn't deterministic (for example when referencing columns from a native query, one of the previous stages, etc.):

```js
[ "field", "CREATED_AT', { "base-type": "type/DateTime" } ]
```

If we want _only_ specific field references, we should update the type-guard to check the ID type
If we're okay with both kinds of references, we should do something else here

**Pull request merged**